udhcpc6 RFC 5970 network-boot options and related DHCPv6 options

Copyright (C) 2026 Tao Li <tao.li06@sap.com>

SPDX-License-Identifier:     GPL-2.0

Adds DHCPv6 option support to udhcpc6:

* RFC 5970 network boot options: BOOT_FILE_URL (59, exported as
  bootfile=), BOOT_FILE_PARAM (60, bootfile_param=), CLIENT_ARCH_TYPE
  (61, client_arch_type=) and NII (62, nii=). BOOT_FILE_PARAM also
  tolerates the dnsmasq length-prefix quirk.
* Common options: DNS_SERVERS (23, dns=), DOMAIN_LIST (24, search=),
  CLIENT_FQDN (39, hostname=). The latter two require and are guarded
  by ENABLE_FEATURE_UDHCP_RFC3397 (dname_dec).
* VENDOR_CLASS/VENDOR_OPTS decoding (vendor_class_id/_data,
  vendor_specific_id/_data, vendoropts=).
* -V/--vendorclass CLI option: sends DHCPv6 option 16 with a numeric
  enterprise number; expected format '<enterprise>:<string>'
  (ONIE passes '42623:onie_vendor:<platform>').
* -O requested options are now actually sent: add_d6_client_options
  builds the ORO from the opt_mask and appends -x options, with bounds
  checks against the on-stack packet buffer (options that do not fit
  are skipped with an error message).
* A caller-supplied client-id (-x 1:...) overrides the generated
  DUID, allowing ONIE to send a DUID-EN.

Known limitation (pre-existing): option_to_env() stops parsing at any
option whose code or length exceeds 255 (checks only the low bytes);
a BOOT_FILE_URL longer than 255 bytes in a packet terminates parsing
of the remaining options.

Must be applied after udhcp6-bugfixes.patch.

diff --git a/networking/udhcp/d6_common.h b/networking/udhcp/d6_common.h
index eb211ea0f..281b54ef2 100644
--- a/networking/udhcp/d6_common.h
+++ b/networking/udhcp/d6_common.h
@@ -80,10 +80,19 @@ struct d6_option {
 #define D6_OPT_INTERFACE_ID  18
 #define D6_OPT_RECONF_MSG    19
 #define D6_OPT_RECONF_ACCEPT 20
+#define D6_OPT_DNS_SERVERS   23
+#define D6_OPT_DOMAIN_LIST   24
+#define D6_OPT_CLIENT_FQDN   39
 
 #define D6_OPT_IA_PD         25
 #define D6_OPT_IAPREFIX      26
 
+/* RFC 5970 */
+#define D6_OPT_BOOT_FILE_URL    59
+#define D6_OPT_BOOT_FILE_PARAM  60
+#define D6_OPT_CLIENT_ARCH_TYPE 61
+#define D6_OPT_NII              62
+
 /*** Other shared functions ***/
 
 struct client6_data_t {
diff --git a/networking/udhcp/d6_dhcpc.c b/networking/udhcp/d6_dhcpc.c
index bc3a5579f..46aa0d512 100644
--- a/networking/udhcp/d6_dhcpc.c
+++ b/networking/udhcp/d6_dhcpc.c
@@ -53,6 +53,7 @@ static const char udhcpc6_longopts[] ALIGN1 =
 	"syslog\0"         No_argument       "S"
 	"request-option\0" Required_argument "O"
 	"no-default-options\0" No_argument   "o"
+	"vendorclass\0"    Required_argument "V"
 	"foreground\0"     No_argument       "f"
 	"background\0"     No_argument       "b"
 ///	IF_FEATURE_UDHCPC_ARPING("arping\0"	No_argument       "a")
@@ -75,9 +76,10 @@ enum {
 	OPT_O = 1 << 11,
 	OPT_o = 1 << 12,
 	OPT_x = 1 << 13,
-	OPT_f = 1 << 14,
+	OPT_V = 1 << 14,
+	OPT_f = 1 << 15,
 /* The rest has variable bit positions, need to be clever */
-	OPTBIT_f = 14,
+	OPTBIT_f = 15,
 	USE_FOR_MMU(             OPTBIT_b,)
 	///IF_FEATURE_UDHCPC_ARPING(OPTBIT_a,)
 	IF_FEATURE_UDHCP_PORT(   OPTBIT_P,)
@@ -143,6 +145,7 @@ static void option_to_env(uint8_t *option, uint8_t *option_end)
 {
 	/* "length minus 4" */
 	int len_m4 = option_end - option - 4;
+
 	while (len_m4 >= 0) {
 		uint32_t v32;
 		char ipv6str[sizeof("ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff")];
@@ -157,7 +160,6 @@ static void option_to_env(uint8_t *option, uint8_t *option_end)
 		case D6_OPT_IA_PD:
 			option_to_env(option + 16, option + 4 + option[3]);
 			break;
-		//case D6_OPT_IA_TA:
 		case D6_OPT_IAADDR:
 /*   0                   1                   2                   3
  *  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
@@ -180,7 +182,47 @@ static void option_to_env(uint8_t *option, uint8_t *option_end)
 			move_from_unaligned32(v32, option + 4 + 16 + 4);
 			*new_env() = xasprintf("lease=%u", (unsigned)v32);
 			break;
-
+		case D6_OPT_DNS_SERVERS: {
+			char *dlist;
+			int off = 4;
+			int len = option[3];
+			dlist = NULL;
+			while (len >= 16) {
+				char *new;
+				sprint_nip6(ipv6str, option + off);
+				new = xasprintf("%s%s%s", dlist ? dlist : "", dlist ? " " : "", ipv6str);
+				free(dlist);
+				dlist = new;
+				off += 16;
+				len -= 16;
+			}
+			if (dlist) {
+				*new_env() = xasprintf("dns=%s", dlist);
+				free(dlist);
+			}
+			break;
+		}
+#if ENABLE_FEATURE_UDHCP_RFC3397
+		case D6_OPT_DOMAIN_LIST: {
+			char *dlist = dname_dec(option + 4, option[3], "");
+			if (dlist) {
+				*new_env() = xasprintf("search=%s", dlist);
+				free(dlist);
+			}
+			break;
+		}
+		case D6_OPT_CLIENT_FQDN: {
+			/* Domain Name starts at option[4+1] */
+			if (option[3] > 1) {
+				char *dlist = dname_dec(option + 5, option[3] - 1, "");
+				if (dlist) {
+					*new_env() = xasprintf("hostname=%s", dlist);
+					free(dlist);
+				}
+			}
+			break;
+		}
+#endif
 		//case D6_OPT_ORO:
 		//case D6_OPT_PREFERENCE:
 		//case D6_OPT_ELAPSED_TIME:
@@ -220,6 +262,56 @@ static void option_to_env(uint8_t *option, uint8_t *option_end)
 			sprint_nip6(ipv6str, option + 4 + 4 + 1);
 			*new_env() = xasprintf("ipv6prefix=%s/%u", ipv6str, (unsigned)(option[4 + 4]));
 			break;
+		case D6_OPT_VENDOR_CLASS:
+		case D6_OPT_VENDOR_OPTS: {
+			uint32_t enterprise;
+			if (option[3] < 4)
+				break;
+			enterprise = (option[4] << 24) | (option[5] << 16) | (option[6] << 8) | option[7];
+			*new_env() = xasprintf("%s_id=%u",
+				option[1] == D6_OPT_VENDOR_CLASS ? "vendor_class" : "vendor_specific",
+				enterprise);
+			if (option[3] > 4) {
+				char *s = xmalloc((option[3] - 4) * 2 + 1);
+				*bin2hex(s, (char*)option + 8, option[3] - 4) = '\0';
+				*new_env() = xasprintf("%s_data=%s",
+					option[1] == D6_OPT_VENDOR_CLASS ? "vendor_class" : "vendor_specific",
+					s);
+				if (option[1] == D6_OPT_VENDOR_OPTS)
+					*new_env() = xasprintf("vendoropts=%s", s);
+				free(s);
+			}
+			break;
+		}
+		case D6_OPT_BOOT_FILE_URL:
+		case D6_OPT_BOOT_FILE_PARAM:
+			{
+				int offset = 0;
+				/* dnsmasq may send length prefix for bootfile-param */
+				if (option[1] == D6_OPT_BOOT_FILE_PARAM &&
+				    option[3] >= 2 &&
+				    option[4] == 0 && option[5] == option[3] - 2)
+					offset = 2;
+
+				*new_env() = xasprintf("%s=%.*s",
+					option[1] == D6_OPT_BOOT_FILE_URL ? "bootfile" : "bootfile_param",
+					(int)(option[3] - offset), (char*)option + 4 + offset);
+			}
+			break;
+		case D6_OPT_CLIENT_ARCH_TYPE: {
+			char *s = xmalloc(option[3] * 2 + 1);
+			*bin2hex(s, (char*)option + 4, option[3]) = '\0';
+			*new_env() = xasprintf("client_arch_type=%s", s);
+			free(s);
+			break;
+		}
+		case D6_OPT_NII: {
+			if (option[3] >= 3) {
+				*new_env() = xasprintf("nii=%02x%02x%02x",
+					option[4], option[5], option[6]);
+			}
+			break;
+		}
 		}
 		option += 4 + option[3];
 		len_m4 -= 4 + option[3];
@@ -280,33 +372,92 @@ static ALWAYS_INLINE uint32_t random_xid(void)
 	return t;
 }
 
+static struct option_set *find_user_clientid_option(void)
+{
+	return udhcp_find_option(client_config.options, D6_OPT_CLIENTID);
+}
+
 /* Initialize the packet with the proper defaults */
 static uint8_t *init_d6_packet(struct d6_packet *packet, char type, uint32_t xid)
 {
 	struct d6_option *clientid;
+	struct option_set *user_clientid;
 
 	memset(packet, 0, sizeof(*packet));
 
 	packet->d6_xid32 = xid;
 	packet->d6_msg_type = type;
 
+	user_clientid = find_user_clientid_option();
+	if (user_clientid && user_clientid->data[OPT_LEN] >= 2) {
+		return packet->d6_options;
+	}
+	if (user_clientid)
+		bb_error_msg("ignoring malformed DHCPv6 client-id from -x 1 (len=%u), falling back to default", user_clientid->data[OPT_LEN]);
+
 	clientid = (void*)client_config.clientid;
 	return d6_store_blob(packet->d6_options, clientid, clientid->len + 2+2);
 }
 
-static uint8_t *add_d6_client_options(uint8_t *ptr)
+static uint8_t *add_d6_client_options(struct d6_packet *packet, uint8_t *ptr)
 {
-	return ptr;
-	//uint8_t c;
-	//int i, end, len;
+	struct option_set *curr;
+	unsigned i, len;
+	uint8_t *oro_ptr = ptr;
+	uint8_t *end = packet->d6_options + sizeof(packet->d6_options);
+	uint16_t *p16;
+
+	/* Add "param req" option (ORO) */
+	if (end - ptr < 4)
+		return ptr;
+	oro_ptr[0] = 0; oro_ptr[1] = D6_OPT_ORO;
+	ptr += 4;
+	p16 = (uint16_t*)ptr;
+	for (i = 0; i < 256; i++) {
+		if (end - (uint8_t*)p16 < 2)
+			break;
+		if (client_config.opt_mask[i >> 3] & (1 << (i & 7))) {
+			*p16++ = htons(i);
+		}
+	}
+	len = (uint8_t*)p16 - (oro_ptr + 4);
+	if (len) {
+		oro_ptr[2] = (len >> 8);
+		oro_ptr[3] = len;
+		ptr = (uint8_t*)p16;
+	} else {
+		ptr = oro_ptr;
+	}
 
-	/* Add a "param req" option with the list of options we'd like to have
-	 * from stubborn DHCP servers. Pull the data from the struct in common.c.
-	 * No bounds checking because it goes towards the head of the packet. */
-	//...
+	/* Add -V vendor class if any */
+	if (client_config.vendorclass) {
+		len = (client_config.vendorclass[2] << 8) | client_config.vendorclass[3];
+		if ((unsigned)(end - ptr) >= len + 4) {
+			memcpy(ptr, client_config.vendorclass, len + 4);
+			ptr += len + 4;
+		} else {
+			bb_error_msg("DHCPv6 vendor class option does not fit in packet, skipping");
+		}
+	}
 
 	/* Add -x options if any */
-	//...
+	curr = client_config.options;
+	while (curr) {
+		len = curr->data[OPT_LEN];
+		if ((unsigned)(end - ptr) >= len + 4) {
+			ptr[0] = 0;
+			ptr[1] = curr->data[OPT_CODE];
+			ptr[2] = 0;
+			ptr[3] = len;
+			memcpy(ptr + 4, curr->data + OPT_DATA, len);
+			ptr += 4 + len;
+		} else {
+			bb_error_msg("DHCPv6 option %u does not fit in packet, skipping", curr->data[OPT_CODE]);
+		}
+		curr = curr->next;
+	}
+
+	return ptr;
 }
 
 static int d6_find_ll_ip(const char *ifname, struct in6_addr *ll_ip)
@@ -361,6 +512,7 @@ static int d6_mcast_from_client_config_ifindex(struct d6_packet *packet, uint8_t
 	);
 }
 
+
 /* Milticast a DHCPv6 Solicit packet to the network, with an optionally requested IP.
  *
  * RFC 3315 17.1.1. Creation of Solicit Messages
@@ -470,7 +622,7 @@ static NOINLINE int send_d6_discover(uint32_t xid, struct in6_addr *requested_ip
 	/* Add options:
 	 * "param req" option according to -O, options specified with -x
 	 */
-	opt_ptr = add_d6_client_options(opt_ptr);
+	opt_ptr = add_d6_client_options(&packet, opt_ptr);
 
 	bb_error_msg("sending %s", "discover");
 	return d6_mcast_from_client_config_ifindex(&packet, opt_ptr);
@@ -523,7 +675,7 @@ static NOINLINE int send_d6_select(uint32_t xid)
 	/* Add options:
 	 * "param req" option according to -O, options specified with -x
 	 */
-	opt_ptr = add_d6_client_options(opt_ptr);
+	opt_ptr = add_d6_client_options(&packet, opt_ptr);
 
 	bb_error_msg("sending %s", "select");
 	return d6_mcast_from_client_config_ifindex(&packet, opt_ptr);
@@ -592,7 +744,7 @@ static NOINLINE int send_d6_renew(uint32_t xid, struct in6_addr *server_ipv6, st
 	/* Add options:
 	 * "param req" option according to -O, options specified with -x
 	 */
-	opt_ptr = add_d6_client_options(opt_ptr);
+	opt_ptr = add_d6_client_options(&packet, opt_ptr);
 
 	bb_error_msg("sending %s", "renew");
 	if (server_ipv6)
@@ -875,13 +1027,14 @@ static void client_background(void)
 //usage:# define IF_UDHCP_VERBOSE(...)
 //usage:#endif
 //usage:#define udhcpc6_trivial_usage
-//usage:       "[-fbnq"IF_UDHCP_VERBOSE("v")"oR] [-i IFACE] [-r IP] [-s PROG] [-p PIDFILE]\n"
+//usage:     "[-fbnq"IF_UDHCP_VERBOSE("v")"oR] [-i IFACE] [-V VENDOR] [-r IP] [-s PROG] [-p PIDFILE]\n"
 //usage:       "	[-x OPT:VAL]... [-O OPT]..." IF_FEATURE_UDHCP_PORT(" [-P N]")
 //usage:#define udhcpc6_full_usage "\n"
 //usage:	IF_LONG_OPTS(
 //usage:     "\n	-i,--interface IFACE	Interface to use (default eth0)"
 //usage:     "\n	-p,--pidfile FILE	Create pidfile"
 //usage:     "\n	-s,--script PROG	Run PROG at DHCP events (default "CONFIG_UDHCPC_DEFAULT_SCRIPT")"
+//usage:     "\n	-V,--vendorclass VENDOR	Vendor identifier"
 //usage:     "\n	-B,--broadcast		Request broadcast replies"
 //usage:     "\n	-t,--retries N		Send up to N discover packets"
 //usage:     "\n	-T,--timeout N		Pause between packets (default 3 seconds)"
@@ -916,6 +1069,7 @@ static void client_background(void)
 //usage:     "\n	-i IFACE	Interface to use (default eth0)"
 //usage:     "\n	-p FILE		Create pidfile"
 //usage:     "\n	-s PROG		Run PROG at DHCP events (default "CONFIG_UDHCPC_DEFAULT_SCRIPT")"
+//usage:     "\n	-V VENDOR	Vendor identifier"
 //usage:     "\n	-B		Request broadcast replies"
 //usage:     "\n	-t N		Send up to N discover packets"
 //usage:     "\n	-T N		Pause between packets (default 3 seconds)"
@@ -950,11 +1104,44 @@ static void client_background(void)
 //usage:     "\n	USR1	Renew lease"
 //usage:     "\n	USR2	Release lease"
 
+static void d6_alloc_vendor_option(const char *str)
+{
+	unsigned len;
+	uint8_t *opt;
+	uint32_t enterprise = 0;
+	char *colon = strchr(str, ':');
+
+	if (colon) {
+		char *end;
+		enterprise = strtoul(str, &end, 0);
+		if (end == colon) {
+			str = colon + 1;
+		} else {
+			enterprise = 0;
+		}
+	}
+
+	len = strlen(str);
+	/* Option 16: code(2) + len(2) + enterprise(4) + [chunk_len(2) + data] */
+	opt = xzalloc(4 + 4 + 2 + len);
+	opt[1] = D6_OPT_VENDOR_CLASS;
+	opt[2] = 0;
+	opt[3] = 4 + 2 + len;
+	opt[4] = (enterprise >> 24);
+	opt[5] = (enterprise >> 16);
+	opt[6] = (enterprise >> 8);
+	opt[7] = enterprise;
+	opt[8] = 0;
+	opt[9] = len;
+	memcpy(opt + 10, str, len);
+	client_config.vendorclass = opt;
+}
 
 int udhcpc6_main(int argc, char **argv) MAIN_EXTERNALLY_VISIBLE;
 int udhcpc6_main(int argc UNUSED_PARAM, char **argv)
 {
 	const char *str_r;
+	const char *str_V;
 	IF_FEATURE_UDHCP_PORT(char *str_P;)
 	void *clientid_mac_ptr;
 	llist_t *list_O = NULL;
@@ -986,19 +1173,22 @@ int udhcpc6_main(int argc UNUSED_PARAM, char **argv)
 	/* O,x: list; -T,-t,-A take numeric param */
 	opt_complementary = "O::x::T+:t+:A+" IF_UDHCP_VERBOSE(":vv") ;
 	IF_LONG_OPTS(applet_long_options = udhcpc6_longopts;)
-	opt = getopt32(argv, "i:np:qRr:s:T:t:SA:O:ox:f"
+	opt = getopt32(argv, "i:np:qRr:s:T:t:SA:O:ox:V:f"
 		USE_FOR_MMU("b")
 		///IF_FEATURE_UDHCPC_ARPING("a")
 		IF_FEATURE_UDHCP_PORT("P:")
 		"v"
-		, &client_config.interface, &client_config.pidfile, &str_r /* i,p */
+		, &client_config.interface, &client_config.pidfile, &str_r /* i,p,r */
 		, &client_config.script /* s */
 		, &discover_timeout, &discover_retries, &tryagain_timeout /* T,t,A */
 		, &list_O
 		, &list_x
+		, &str_V
 		IF_FEATURE_UDHCP_PORT(, &str_P)
 		IF_UDHCP_VERBOSE(, &dhcp_verbose)
 	);
+	if (opt & OPT_V)
+		d6_alloc_vendor_option(str_V);
 	requested_ipv6 = NULL;
 	if (opt & OPT_r) {
 		if (inet_pton(AF_INET6, str_r, &ipv6_buf) <= 0)
